dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -1238,44 +1390,128 @@ RtpsUdpDataLink::WriterInfo::should_nack() const
 }
 
 bool
-RtpsUdpDataLink::RtpsReader::nack_durable(const WriterInfo& info)
+RtpsUdpDataLink::RtpsWriter::add_reader(const RepoId& id, const ReaderInfo& info)
 {
-  return durable_ && (info.recvd_.empty() ||
-                      info.recvd_.low() > info.hb_range_.first);
+  bool result = false;
+  ACE_GUARD_RETURN(ACE_Thread_Mutex, g, mutex_, false);
+  ReaderInfoMap::const_iterator iter = remote_readers_.find(id);
+  if (iter == remote_readers_.end()) {
+    remote_readers_.insert(ReaderInfoMap::value_type(id, info));
+    ready_to_hb_ = !info.durable_;
+    result = true;
+  }
+  return result;
 }
 
-void
-RtpsUdpDataLink::send_ack_nacks(RtpsReaderMap::value_type& rr, bool finalFlag)
+bool
+RtpsUdpDataLink::RtpsWriter::has_reader(const RepoId& id) const
 {
-  using namespace OpenDDS::RTPS;
+  ACE_GUARD_RETURN(ACE_Thread_Mutex, g, mutex_, false);
+  return remote_readers_.find(id) != remote_readers_.end();
```

Comment:
```suggestion
  return remote_readers_.count(id);
```

Reply:
I'm not really a fan of implicit integer / pointer -> boolean conversions. Having the explicit comparison keeps the types as you'd expect and causes helpful compiler warnings if/when the surrounding code changes, in addition to (usually) being optimized away by the compiler. I get that something like this might make the code-as-written a little more concise & readable, but I'm not sure the resilience / maintainability tradeoff is worth it. In addition, if remote_readers_ ever needed to turn into a multiset (for whatever reason), this would be doing more work than is strictly required by finding all the matches and not just _any_ match.

Reply:
You can add the `!= 0` if you want, the more important part is that a `count` operation is different than a `find` operation.